server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,292 @@
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
+
+    for the future:
+    Find combination of potential games that allows for maximal number of games to launch
+    Repeat 8. with an increasing amount of top games removed beforehand to get all possible game combinations
+    If there is more than one solution, pick the one with highest total game quality
+    Optimization: sort by game quality first, abort when you find a solution with
+    the full number of theoretically possible games (floor(playersInQueue/(teamSize*2)))
+    """
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        self._logger.debug("=== starting matching algorithm ===")
+
+        searches = SortedList(searches, key=lambda s: s.average_rating)
+        possible_games = set()
+        for index, search in enumerate(searches):
+
+            self._logger.debug("building game for %s", repr(search))
+
+            try:
+                participants = self._pick_neighboring_players(searches, index)
+                match = self.make_teams(participants)
+                game = self.calculate_game_quality(match)
+                possible_games.add(game)
+            except NotEnoughPlayersException:
+                self._logger.warning("Couldn't pick enough players for a full game. Skipping this game...")
+            except UnevenTeamsException:
+                self._logger.warning("Failed to assign even teams. Skipping this game...")
+
+        self._logger.debug("got %i games", len(possible_games))
+        for game in possible_games:
+            self._logger.debug("game: %s vs %s rating disparity: %i quality: %f",
+                               repr(game.match[0]),
+                               repr(game.match[1]),
+                               game.match[0].cumulated_rating - game.match[1].cumulated_rating,
+                               game.quality)
+
+        return self._pick_best_noncolliding_games(list(possible_games))
+
+    def _pick_neighboring_players(self, searches: List[Search], index: int) -> List[Search]:
+        """
+        Picks searches from the list starting with the search at the given index and then expanding in both directions
+        until there are enough players for a full game.
+        Raises NotEnoughPlayersException if it can't find enough suitable searches to fill a game.
+        """
+        participants = []
+        i = 0
+        number_of_players = 0
+        lower = searches[:index]
+        lower = iter(lower[::-1])
+        higher = iter(searches[index:])
+
+        failed_last_time = False
+        while number_of_players < self.team_size * 2:
+            candidate = next(lower if i % 2 else higher, None)
```

Comment:
You could do something like this:
```suggestion
            candidate, prev = next(lower if i % 2 else higher, None), candidate
```
Then your break condition might be simpler:
```python3
if candidate is None:
    if prev is None:
        raise NotEnoughPlayersException
    continue
```